src/lando/utils/github.py

```diff
@@ -1,36 +1,79 @@
 import asyncio
 import logging
+import re
 
 import requests
 from django.conf import settings
 from simple_github import AppAuth, AppInstallationAuth
 
-from lando.main.models.repo import Repo
+from lando.utils.const import URL_USERINFO_RE
 
 logger = logging.getLogger(__name__)
 
 
-class GitHubAPI:
-    """A simple wrapper that authenticates with and communicates with the GitHub API."""
+class GitHub:
+    """Work with authentication to GitHub repositories."""
 
-    GITHUB_BASE_URL = "https://api.github.com"
+    GITHUB_URL_RE = re.compile(
+        f"https://{URL_USERINFO_RE.pattern}?github.com/(?P<owner>[-A-Za-z0-9]+)/(?P<repo>[^/]+)(.git)?/?"
+    )
 
-    def __init__(self, repo: Repo):
-        repo_owner = repo._github_repo_org
-        repo_name = repo.git_repo_name
+    repo_url: str
+    repo_owner: str
+    repo_name: str
+    userinfo: str
 
-        self.session = requests.Session()
-        self.session.headers.update(
-            {
-                "Authorization": f"Bearer {self._get_token(repo_owner, repo_name)}",
-                "User-Agent": settings.HTTP_USER_AGENT,
-                "Accept": "application/vnd.github+json",
-                "X-GitHub-Api-Version": "2022-11-28",
-            }
+    def __init__(self, repo_url: str):
+        # GitHub doesn't allow '.git' to be at the end of a repository name, so we
+        # normalise it out to be sure. This include first removing any trailing /
+        self.repo_url = repo_url.removesuffix("/").removesuffix(".git")
+
+        parsed_url = self.parse_url(self.repo_url)
+
+        if parsed_url is None:
+            raise ValueError(f"Cannot parse URL as GitHub repo: {repo_url}")
+
+        self.repo_owner = parsed_url["owner"]
+        self.repo_name = parsed_url["repo"]
+        self.userinfo = parsed_url["userinfo"]
+
+    @classmethod
+    def is_supported_url(cls, url: str) -> bool:
+        return cls.parse_url(url) is not None
+
+    @classmethod
+    def parse_url(cls, url: str) -> re.Match[str] | None:
+        """Parse GitHub data from URL, or return None if not Github.
+
+        Note: no normalisation is performed on the URL
+        """
+        return re.match(cls.GITHUB_URL_RE, url)
+
+    @property
+    def authenticated_url(self) -> str:
+        """Return an authenticated URL, suitable for use with `git` to push and pull.
+
+        Note: any '.git' or trailing '/' will be normalised out.
```

Comment:
Not sure this line is relevant to this method, as it isn't doing any of these things.